src/replication.c

```diff
@@ -3390,3 +3416,192 @@ void replicationStartPendingFork(void) {
+
+        server.target_replica_host = zstrdup(c->argv[1]->ptr);
+        server.target_replica_port = port;
+    }
+
+    /* Check for optional FORCE argument, ignore invalid argument */
+    if ((c->argc > 4) && !strcasecmp(c->argv[4]->ptr, "force")) {
+        if (server.target_replica_host) {
+            server.force_failover = 1;
+        } else {
+            addReplyError(c,"FAILOVERTO cannot failover to \"any one\" with "
+                            "force flag.");
+            return;
+        }
+    }
+
+    mstime_t now = mstime();
+    server.failover_end_time = now + timeout_in_ms;
+    /* Cluster failover pauses for timeout_in_ms*2 so do it here too, we will
+     * unpause early in failoverCron if failover times out. */
+    pauseClients(now + (timeout_in_ms*2),CLIENT_PAUSE_WRITE);
+
+    serverLog(LL_NOTICE,"FAILOVERTO %s:%s (timeout: %ld ms) requested.",
+        (char*) c->argv[1]->ptr, (char*) c->argv[2]->ptr, timeout_in_ms);
+
+    addReply(c,shared.ok);
+}
+
+/* Failover cron function, checks coordinated failover state. */
+void failoverCron(void) {
+    mstime_t now = mstime();
+
+    /* Check if failover already completed */
+    if (server.masterhost && (server.repl_state >= REPL_STATE_TRANSFER)) {
+        serverLog(LL_NOTICE,"FAILOVERTO %s:%d completed successfully.",
+            server.masterhost, server.masterport);
+        goto cron_complete;
+    }
+
+    /* Check if failover operation has timed out */
+    if (server.failover_end_time <= now) {
+        if (server.masterhost) {
+            /* We attempted to failover but it did not complete in time so
+             * become a master again to undo the failover. */
+            serverLog(LL_NOTICE,
+                "FAILOVERTO %s:%d attempted, but sync was unsuccessful.",
+                server.target_replica_host, server.target_replica_port);
+            replicationUnsetMaster();
+            goto cron_complete;
+        } else if (server.force_failover) {
+            /* If timeout has expired force a failover if requested. */
+            replicationSetMaster(server.target_replica_host,
+                server.target_replica_port);
+            serverLog(LL_NOTICE,
+                "FAILOVERTO %s:%d expired, but forcing failover.",
+                server.target_replica_host, server.target_replica_port);
+            /* Extend failover time to allow replication to complete, then reset
+             * the force flag so progress will be checked during the extended
+             * duration. */
+            server.failover_end_time += FAILOVERTO_TIMEOUT;
+            server.force_failover = 0;
+            return;
+        } else {
+            /* Failover did not attempt or succeed. */
+            serverLog(LL_NOTICE,"FAILOVERTO %s:%d was unsuccessful.",
+                server.target_replica_host, server.target_replica_port);
+            goto cron_complete;
+        }
+    }
+
+    /* If offsets are equal and we havn't started failover yet, do so */
```

Comment:
Maybe I'm missing something, but I don't see the check for if the failover has started yet. There seems to be an implicit check since we should disconnect all of our replicas when we call replicationSetMaster, so we should no longer be able to find a valid replica. However, a replica could theoretically reconnect as a chain replica.

Reply:
added a check for this